src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleMetadata.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Text.Json.Serialization;
+
+namespace System.Text.Json
+{
+    public static partial class JsonSerializer
+    {
+        // Pre-encoded metadata properties.
+        private static readonly JsonEncodedText s_metadataId = JsonEncodedText.Encode("$id", encoder: null);
+        private static readonly JsonEncodedText s_metadataRef = JsonEncodedText.Encode("$ref", encoder: null);
+        private static readonly JsonEncodedText s_metadataValues = JsonEncodedText.Encode("$values", encoder: null);
+
+        internal static MetadataPropertyName WriteReferenceForObject(
```

Comment:
nit: I see many duplicated code on the three callers of this method.

Reply:
The pattern is:
```cs

                    if (options.ReferenceHandling.ShouldWritePreservedReferences())
                    {
                        if (JsonSerializer.WriteReferenceForObject(this, dictionary, ref state, writer) == MetadataPropertyName.Ref)
                        {
                            writer.WriteEndObject();
                            return true;
                        }
                    }
```

For `ShouldWritePreservedReferences` I prefer inline to avoid the overhead of calling a no-op method.

For `writer.WriteEndObject()` I kept the "EndObject" in the caller, but will change it to the method and remove from the three callers.

Reply:
I must have been mistaken. Thanks for clarify.